spring-rabbit-test/src/main/java/org/springframework/amqp/rabbit/test/TestRabbitTemplate.java

```diff
@@ -0,0 +1,200 @@
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.springframework.amqp.core.Message;
+import org.springframework.amqp.core.MessageBuilder;
+import org.springframework.amqp.core.MessageListener;
+import org.springframework.amqp.rabbit.connection.ConnectionFactory;
+import org.springframework.amqp.rabbit.core.ChannelAwareMessageListener;
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+import org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer;
+import org.springframework.amqp.rabbit.listener.RabbitListenerEndpointRegistry;
+import org.springframework.amqp.rabbit.listener.adapter.AbstractAdaptableMessageListener;
+import org.springframework.amqp.rabbit.support.CorrelationData;
+import org.springframework.amqp.rabbit.support.RabbitExceptionTranslator;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.SmartInitializingSingleton;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+
+import com.rabbitmq.client.AMQP.BasicProperties;
+import com.rabbitmq.client.Channel;
+import com.rabbitmq.client.Envelope;
+
+/**
+ * A {@link RabbitTemplate} that invokes {@code @RabbitListener} s directly.
+ * It currently only supports the queue name in the routing key.
+ * It does not currently support publisher confirms/returns.
+ *
+ * @author Gary Russell
+ * @since 2.0
+ *
+ */
+public class TestRabbitTemplate extends RabbitTemplate implements ApplicationContextAware, SmartInitializingSingleton {
+
+	private static final String REPLY_QUEUE = "testRabbitTemplateReplyTo";
+
+	private final Map<String, Listeners> listeners = new HashMap<>();
+
+	private ApplicationContext applicationContext;
+
+	@Autowired
+	private RabbitListenerEndpointRegistry registry;
+
+	public TestRabbitTemplate(ConnectionFactory connectionFactory) {
+		super(connectionFactory);
+		setReplyAddress(REPLY_QUEUE);
+	}
+
+	@Override
+	public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+		this.applicationContext = applicationContext;
+	}
+
+	@Override
+	public void afterSingletonsInstantiated() {
+		this.registry.getListenerContainers()
+			.stream()
+			.map(container -> (AbstractMessageListenerContainer) container)
+			.forEach(c -> {
+				for (String queue : c.getQueueNames()) {
+					setupListener(c, queue);
+				}
+			});
+		this.applicationContext.getBeansOfType(AbstractMessageListenerContainer.class).values()
+			.stream()
+			.forEach(container -> {
+				for (String queue : container.getQueueNames()) {
+					setupListener(container, queue);
+				}
+			});
+	}
+
+	private void setupListener(AbstractMessageListenerContainer container, String queue) {
+		Listeners listenersForQueue = this.listeners.get(queue);
+		if (listenersForQueue == null) {
```

Comment:
@garyrussell , are you going to address this concern?
Thanks

@panchenko , thank you very much for review!